monai/losses/dice.py

```diff
@@ -298,5 +300,177 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
         return f
 
 
+class GeneralizedWassersteinDiceLoss(_Loss):
+    """
+    Generalized Wasserstein Dice Loss [1] in PyTorch.
+    Compared to [1] we used a weighting method similar to the one
+    used in the generalized Dice Loss [2].
+
+    References:
+    ===========
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017.
+    [2] "Generalised dice overlap as a deep learning loss function
+        for highly unbalanced segmentations",
+        Sudre C., et al. MICCAI DLMIA 2017.
+
+    wasserstein_distance_map:
+    Compute the voxel-wise Wasserstein distance (eq. 6 in [1]) between the
+    flattened prediction and the flattened labels (ground_truth) with respect
+    to the distance matrix on the label space M.
+    References:
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017
+
+    compute_weights_generalized_true_positives:
+    Compute the weights \alpha_l of eq. 9 in [1] but using the weighting
+    method proposed in the generalized Dice Loss [2].
+    References:
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017
+    [2] "Generalised dice overlap as a deep learning loss function
+        for highly unbalanced segmentations." Sudre C., et al.
+        MICCAI DLMIA 2017.
+    """
+
+    def __init__(self, dist_matrix, reduction="mean"):
+        """
+        Args:
+            param dist_matrix: 2d tensor or 2d numpy array; matrix of distances
+            between the classes. It must have dimension C x C where C is the
+            number of classes.
+            param reduction: str; reduction mode.
+
+        Raises:
+            ValueError: dist_matrix.shape[0] != dist_matrix.shape[1] is invalid.
+
+        """
+        super(GeneralizedWassersteinDiceLoss, self).__init__(reduction=reduction)
+
+        if dist_matrix.shape[0] != dist_matrix.shape[1]:
+            raise ValueError("Dist Matrix is invalid.")
+
+        self.M = dist_matrix
+        if isinstance(self.M, np.ndarray):
+            self.M = torch.from_numpy(self.M)
+        if torch.cuda.is_available():
+            self.M = self.M.cuda()
```

Comment:
This will move M to device 0 which won't necessarily be compatible with multi-GPU code. What I think we have to do is move M to the correct device in `forward`, in which case you would have a local copy of M constructed by `M = torch.clone(self.M).to(input.get_device())` which is used in place of `self.M` everywhere. I think this is the correct way of doing it although it involves copying M every forward pass, alternatively you could recall what the previous value of `input.get_device()` was and reuse the created M copy if it's same.

Reply:
thats something i wasnt aware of. i will update the code with your suggestion.